src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializerHelpers.cs

```diff
@@ -0,0 +1,364 @@
+            IWebApiUrlHelper urlHelper = readContext.InternalUrlHelper;
+            ODataPath odataPath = null;
+
+            if (internalRequest != null && urlHelper != null)
+            {
+                string serviceRoot = urlHelper.CreateODataLink(
+                    internalRequest.Context.RouteName,
+                    internalRequest.PathHandler,
+                    new List<ODataPathSegment>());
+
+                try
+                {
+                    odataPath = pathHandler.Parse(serviceRoot, id, internalRequest.RequestContainer).Path;
+                }
+                catch (ODataException)
+                {
+                    return null;
+                }
+            }
+
+            return odataPath;
+        }
+
+        internal static void ApplyODataIdContainer(object resource, ODataResourceWrapper resourceWrapper,
+            ODataDeserializerContext readContext)
+        {
+            Routing.ODataPath path = readContext.Path;
+            if (path == null)
+            {
+                return;
+            }
+
+            IEdmEntityType entityType = path.EdmType.AsElementType() as IEdmEntityType;
+
+            if (entityType != null)
+            {
+                //Setting Odataid , for POCO classes, as a property in the POCO object itself(if user has OdataIDContainer property),
+                //for Delta and EdmEntity object setting as an added property ODataIdcontianer in those classes
+
+                ODataPath ODataPath = new ODataPath(path.Segments);
+
+                // if there is no Id on the resource, try to compute one from path
+                if (resourceWrapper.ResourceBase.Id == null)
+                {
+                    ODataUri odataUri = new ODataUri { Path = ODataPath };
+                    resourceWrapper.ResourceBase.Id = odataUri.BuildUri(ODataUrlKeyDelimiter.Parentheses);
+                }
+
+                if (resourceWrapper.ResourceBase.Id != null)
+                {
+                    string odataId = resourceWrapper.ResourceBase.Id.OriginalString;
+
+                    IODataIdContainer container = new ODataIdContainer();
+                    container.ODataId = odataId;
+
+                    if (resource is EdmEntityObject edmObject)
+                    {
+                        edmObject.ODataIdContainer = container;
+                        edmObject.ODataPath = ODataPath;
+                    }
+                    else if (resource is IDeltaSetItem deltasetItem)
+                    {
+                        deltasetItem.ODataIdContainer = container;
+                        deltasetItem.ODataPath = ODataPath;
+                    }
+                    else
+                    {
+                        PropertyInfo containerPropertyInfo = EdmLibHelpers.GetClrType(entityType, readContext.Model).GetProperties().Where(x => x.PropertyType == typeof(IODataIdContainer)).FirstOrDefault();
```

Comment:
If there were more than 1 properties of type `IODataIdContainer` in the POCO class, isn't that reason enough to throw an exception? If true, we should move this logic for retrieving the `IODataContainer` property to a helper method and modify the logic such that we throw an exception if we find such a scenario

Reply:
Interesting case!

Reply:
Am thinking that we should be able to have more than one property of `IODataIdContainer` in the POCO class. I know this code segment doesn't take care of it because it gets  the `FirstOrDefault` property but it should be a valid scenario.. One `POCO` class can have more than nested property and we should be able to use `@odata.id` for all nested properties.